File: 415_add_strings.java
class Solution {
    public String addStrings(String num1, String num2) {
        int len1 = num1.length(), len2 = num2.length();
        int idx1 = len1 - 1, idx2 = len2 - 1;
        int carry = 0;
        StringBuilder sb = new StringBuilder();
        while (idx1 >= 0 && idx2 >= 0) {
            int n1 = Integer.parseInt(String.valueOf(num1.charAt(idx1)));
            int n2 = Integer.parseInt(String.valueOf(num2.charAt(idx2)));
            int sum = n1 + n2 + carry;
            if (sum >= 10) {
                carry = sum / 10;
                sum %= 10;
            } else {
                carry = 0;
            }
            sb.insert(0, sum);
            idx1--;
            idx2--;
        }

        if (idx1 >= 0 && idx2 < 0) {
            while (idx1 >= 0) {
                int n = Integer.parseInt(String.valueOf(num1.charAt(idx1)));
                int sum = n + carry;
                if (sum >= 10) {
                    carry = sum / 10;
                    sum %= 10;
                } else { carry = 0; }
                sb.insert(0, sum);
                idx1--;
            }
        } else if (idx1 < 0 && idx2 >= 0) {
            while (idx2 >= 0) {
                int n = Integer.parseInt(String.valueOf(num2.charAt(idx2)));
                int sum = n + carry;
                if (sum >= 10) {
                    carry = sum / 10;
                    sum %= 10;
                } else { carry = 0; }
                sb.insert(0, sum);
                idx2--;
            }
        }
        if (carry > 0) {
            sb.insert(0, carry);
        }

        return sb.toString();
    }
}

File: python/415_add_strings.py
class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        num1_idx = len(num1) - 1
        num2_idx = len(num2) - 1

        carry = 0
        res = []
        while num1_idx >= 0 or num2_idx >= 0:
            n1 = 0
            if num1_idx >= 0:
                n1 = (ord(num1[num1_idx]) - ord('0'))
                num1_idx -= 1
            n2 = 0
            if num2_idx >= 0:
                n2 = (ord(num2[num2_idx]) - ord('0'))
                num2_idx -= 1

            accum = carry + n1 + n2
            carry = accum // 10
            accum = accum % 10
            res.insert(0, str(accum))

        if carry > 0:
            res.insert(0, str(carry))
        return "".join(res)

